File: examples/typescript/followup-digest.ts
/**
 * Nondual — Daily followup digest
 *
 * Run this on a schedule (cron, GitHub Actions, etc.) to get a
 * daily digest of what needs attention. Checks what's due today,
 * groups by company, and outputs a prioritised action list.
 *
 * Usage:
 *   NONDUAL_API_KEY=nd_... npx tsx followup-digest.ts
 *
 * Cron (daily at 8am):
 *   0 8 * * * NONDUAL_API_KEY=nd_... npx tsx followup-digest.ts
 */
import { Nondual } from 'nondual';

const nd = new Nondual({ agent: 'digest-agent' });

async function main() {
  const today = new Date().toISOString().slice(0, 10);
  console.log(`=== Followup digest for ${today} ===\n`);

  const res = await nd.listOpenFollowups({ due_before: today }) as any;
  const followups = res.followups ?? [];

  if (followups.length === 0) {
    console.log('Nothing due today. ✅');
    return;
  }

  console.log(`${followups.length} open followup(s):\n`);

  // Group by company domain
  const byCompany = new Map<string, typeof followups>();
  for (const f of followups) {
    const domain = f.contact?.company?.domain ?? 'no company';
    if (!byCompany.has(domain)) byCompany.set(domain, []);
    byCompany.get(domain)!.push(f);
  }

  for (const [domain, items] of byCompany) {
    console.log(`── ${domain} ──`);
    for (const f of items) {
      const c = f.contact;
      const dnd = c?.do_not_disturb ? ' ⛔ DND' : '';
      const name = c?.name ?? c?.identifiers?.emails?.[0] ?? '?';
      const due = f.due ? f.due.slice(0, 10) : 'no date';
      const overdue = f.due && f.due < today ? ' ⚠️ OVERDUE' : '';
      console.log(`  ${name}${dnd}${overdue}`);
      console.log(`  → ${f.action} (due ${due})`);
      if (f.owner) console.log(`     owner: ${f.owner}`);
      console.log();
    }
  }

  // Overdue summary
  const overdue = followups.filter((f: any) => f.due && f.due < today);
  if (overdue.length > 0) {
    console.log(`⚠️  ${overdue.length} overdue — act today.`);
  }
}

main().catch(console.error);

File: examples/python/followup-digest.py
"""
Nondual — Daily followup digest (Python)

Fetches today's open followups and prints a prioritised digest.
Run on a schedule (cron, GitHub Actions, Hermes cron, etc.)

Usage:
  pip install requests
  NONDUAL_API_KEY=nd_... python followup-digest.py

Cron (daily 8am):
  0 8 * * * NONDUAL_API_KEY=nd_... python /path/to/followup-digest.py
"""

import os
import sys
import requests
from datetime import date
from collections import defaultdict

API = "https://api.nondual.cloud/v1"
KEY = os.environ.get("NONDUAL_API_KEY", "")
HEADERS = {"Authorization": f"Bearer {KEY}", "Content-Type": "application/json"}

today = date.today().isoformat()


def get_followups(due_before=today):
    r = requests.get(
        f"{API}/list-open-followups",
        headers=HEADERS,
        params={"due_before": due_before},
        timeout=15,
    )
    r.raise_for_status()
    return r.json().get("followups", [])


def main():
    if not KEY:
        print("Error: NONDUAL_API_KEY not set")
        sys.exit(1)

    followups = get_followups()

    print(f"=== Followup digest — {today} ===\n")

    if not followups:
        print("Nothing due today. ✅")
        return

    print(f"{len(followups)} open followup(s):\n")

    # Group by company domain
    by_company = defaultdict(list)
    for f in followups:
        domain = (f.get("contact") or {}).get("company", {}).get("domain") or "no company"
        by_company[domain].append(f)

    overdue = [f for f in followups if f.get("due") and f["due"] < today]

    for domain, items in sorted(by_company.items()):
        print(f"── {domain} ──")
        for f in items:
            c = f.get("contact") or {}
            name = c.get("name") or (c.get("identifiers") or {}).get("emails", ["?"])[0]
            dnd = " ⛔ DND" if c.get("do_not_disturb") else ""
            due_str = (f.get("due") or "no date")[:10]
            late = " ⚠️ OVERDUE" if f.get("due") and f["due"] < today else ""
            owner = f"  owner: {f['owner']}" if f.get("owner") else ""
            print(f"  {name}{dnd}{late}")
            print(f"  → {f['action']} (due {due_str}){owner}")
            print()

    if overdue:
        print(f"⚠️  {len(overdue)} overdue — act today.")


if __name__ == "__main__":
    main()
